### ISSUE TITLE
Add a Project scope

### ISSUE THREAD

**Issue by nikku** [camunda/camunda-modeler#2508]
__What should we do?__

Add the infrastructure for the Camunda Modeler to recognize not only single BPMN files but _projects_ and appropriately represent them.

* [ ] Be able to group open diagrams / folders into a project (https://github.com/camunda/camunda-modeler/issues/1351)
* [ ] Be able to switch between projects
* [ ] Be able to see which files / folders are opened in a project via a side-bar
    * [ ] Also non-BPMN files are represented, clicking them will open in the IDE associated with them ("play nicely with others")
* [ ] Make project context accessible to infrastructure so that we can build on top of it
    * [ ] Collecting of meta-data (i.e. used variables)
    * [ ] Discover related diagrams and link between them (https://github.com/camunda/camunda-modeler/issues/112)

__Why should we do it?__

This is an essential feature for advanced text editors. Without it we won't be able to reliably implement advanced tooling and linking of diagrams.

**Comment by philippfromme** [comment]
# Architecture Discussion

I (@philippfromme) discussed with @barmac.

## Key Points

- Process applications feature should be independent of React
- Process application files will be managed separately from tabs, rather than as a new supported file type that isn't opened in a tab
- No decision reached on supporting files only or folders

## File vs. Folder Support

### Files

**Pros:**
- Easier to implement (single file reading and writing already implemented)
- Flat list of files more closely resembles process applications in Web Modeler
- Potentially easier to use in both modelers

**Cons:**

- Deleting, renaming, or moving a file will break the process application

### Folders (reading BPMN, DMN, and Form files only)

**Pros:**

- Adding a single folder automatically picks up all files within
- No need to explicitly add files
- Deleting, renaming, or moving a file will not break the process application

**Cons:**

- Requires folder reading functionality, which is currently not implemented

## Next Steps

1. Create a diagram to visualize the proposed architecture at the C4 components level
2. Make a decision on whether to support files only or folders

@nikku I would appreciate your input.

**Comment by nikku** [comment]
I strongly support "folder reading support". We don't want to oversimplify our process applications feature, but support the full power of what pro-coders do:

* Resources relevant to the application may be arbitrarily nested _within_ a certain root folder
* They may be moved, renamed, etc. and the app must react to that.
* In simple terms this can be accomplished by not only watching files (we already do this), but by watching a folder; we must watch it anyway in order to do smart background indexing (to derive intelligence)

**Comment by philippfromme** [comment]
Okay, we'll implement folders then. I'm also in favor of folders as I think with files only we'll end up wishing we'd implemented folders. 🤔

**Comment by lmbateman** [comment]
> Deleting, renaming, or moving a file will not break the process application

Is that because we create a pointer to each file, or for some other reason? In other words, is the process application only aware of the files in that folder, or does it track files even after they're moved out of the folder?

**Comment by philippfromme** [comment]
> Is that because we create a pointer to each file, or for some other reason? In other words, is the process application only aware of the files in that folder, or does it track files even after they're moved out of the folder?

The process application tracks any relevant files (BPMN, DMN and Form) in the root directory or any of it's child directories. Moving a file outside of that root directory will effectively remove the file from the process application.

**Comment by lmbateman** [comment]
> Deleting, renaming, or moving a file will not break the process application
> Moving a file outside of that root directory will effectively remove the file from the process application.

If the moved/deleted file is a linked resource, wouldn't that break the process application? Or is the first sentence above only referring to BPMN files?

**Comment by nikku** [comment]
> If the moved/deleted file is a linked resource, wouldn't that break the process application? Or is the first sentence above only referring to BPMN files?

Of course it would break the process application, in similar ways like you fool someone by removing the chair he sits on: If the thing is used ("you sit on the chair") you'll notice that the thing ("the chair") is missing. If you're not sitting on it, then the operation is absolutely fine.

We should build our validation in a robust manner, so that we detect that a piece we criticially rely on is now gone ("someone removed the chair I sit on"). This is out of question. But we should also build the underlying system in a way that it is robust, i.e. _will_ be able to handle the situation where the chair you sit on is gone. 

(Not sure if that makes sense).

**Comment by philippfromme** [comment]
Yes, we would show an error that the linked resource is now missing. 🙂

**Comment by philippfromme** [comment]
@barmac https://github.com/camunda/camunda-modeler/pull/4762 will close all the issues that I've ✅'d. The remaining issues are what we should tackle next. Your work on the [task appending framework](https://github.com/camunda/product-hub/issues/2608) would cover https://github.com/camunda/camunda-modeler/issues/4668.

**Comment by barmac** [comment]
I will reuse this issue in the epic. Thanks for the headsup.

**Comment by barmac** [comment]
Not all of the items scoped in the top comment are shipped with 5.34.0. However, we are delivering the first iteration. @philippfromme perhaps we should update this epic issue so that it does not talk about first iteration anymore?

**Comment by barmac** [comment]
Since this epic goes beyond the scope shipped with 8.7, I am removing it from the milestone.

**Comment by se16n** [comment]
Hello,

I'm wondering what is the current planning for #4676 ?

All BPMN modeling guidelines suggest using business process levels, but currently there's no way to navigate between linked process models back and forth. The problem is even worse because Camunda Modeler is the only BPMN modeling software available for the local installation with an acceptable license.

**Comment by nikku** [comment]
@se16n I assume you're on Camunda 8? Indeed we need to follow-up on navigation; we [do support linking](https://docs.camunda.io/docs/next/components/modeler/desktop-modeler/process-applications/#linking-resources) already, in the context of a process application.

**Comment by se16n** [comment]
> [@se16n](https://github.com/se16n) I assume you're on Camunda 8? Indeed we need to follow-up on navigation; we [do support linking](https://docs.camunda.io/docs/next/components/modeler/desktop-modeler/process-applications/#linking-resources) already, in the context of a process application.

Linking is great, but it doesn't allow you to navigate between linked entities (another BPMN process called via Call Activity is the main use case I guess).
